pulumiyaml/program.go

```diff
@@ -0,0 +1,118 @@
+package pulumiyaml
+
+import "fmt"
+
+type Program struct {
+	// Defaults to "test"
+	Name string `yaml:"name,omitempty"`
+	// Defaults to "yaml"
+	Runtime     string              `yaml:"runtime,omitempty"`
+	Description string              `yaml:"description,omitempty"`
+	Config      map[string]any      `yaml:"config"`
+	Plugins     Plugins             `yaml:"plugins"`
+	Resources   map[string]Resource `yaml:"resources"`
+}
+
+func NewProgram() *Program {
+	return &Program{
+		Name:      "test",
+		Runtime:   "yaml",
+		Config:    make(map[string]any),
+		Resources: make(map[string]Resource),
+	}
+}
+
+// AddResource adds a resource to the program by name.
+// Returns an error if the resource already exists in the program.
+func (p *Program) AddResource(name string, resource Resource) error {
+	if p.Resources == nil {
+		p.Resources = make(map[string]Resource)
+	}
+	if _, ok := p.Resources[name]; ok {
+		return fmt.Errorf("resource %q already exists in program", name)
+	}
+	p.Resources[name] = resource
+	return nil
+}
+
+// RemoveResource removes a resource from the program by name and returns it, if found.
+func (p *Program) RemoveResource(name string) (*Resource, error) {
+	if p.Resources == nil {
+		return nil, fmt.Errorf("resource %q not found in program", name)
+	}
+	removed, found := p.Resources[name]
+	if !found {
+		return nil, fmt.Errorf("resource %q not found in program", name)
+	}
+	delete(p.Resources, name)
+	return &removed, nil
+}
+
+// GetResource returns a resource from the program by name.
+func (p *Program) GetResource(name string) (*Resource, error) {
+	resource, found := p.Resources[name]
+	if !found {
+		return nil, fmt.Errorf("resource %q not found in program", name)
+	}
+	return &resource, nil
+}
+
+// UpdateResource updates a resource in the program by name.
+// Returns the old resource and true if the resource was found and updated.
+func (p *Program) UpdateResource(name string, resource Resource) (*Resource, error) {
+	oldResource, found := p.Resources[name]
+	if !found {
+		return nil, fmt.Errorf("resource %q not found in program", name)
+	}
+	p.Resources[name] = resource
+	return &oldResource, nil
+}
+
+type Resource struct {
```

Comment:
Any reason for not re-using the structs from here?
https://github.com/pulumi/pulumi-yaml/blob/88532b6cf9e6eb492376c272d6ffc00edd1b2a15/pkg/pulumiyaml/template.go#L3

Reply:
Huh ... had looked around but not found this! Will swap it over. If there's anything that's missing, that would be a sensible place to contribute it back.

Reply:
Hmm, just trying this now, it seems like it's designed for a template rather than an actual program. Most notably, it's missing the runtime option - in this context I guess we could just hard-code it to "yaml".

There's also the [`Project` struct in the automation API](https://github.com/pulumi/pulumi/blob/sdk/v3.129.0/sdk/go/common/workspace/project.go#L166) but this seems to be lacking the fields for resources. I also have a feeling that doing any workspace configuration via the Automation API might result in all the resources being delete as it's not currently round-tripable.

Reply:
Perhaps we should contribute the resource fields to the automation API and just use that given that's the library we're wrapping.